### PR TITLE
Add more prometheus metrics.

### PR DESCRIPTION
**Pull request by winder**
## Summary

Added some new metrics to go along with `indexer_daemon_import_time_sec`:
-  block_upload_time_sec
-  imported_tx_per_block
-  imported_round

Also moved the location where metrics are collected, previously they did not represent exactly one round worth of data.

Updated the block generator import test runner to capture the new metrics.

New report output
```
cat cmd/block-generator/results/config.asset.close.report
test_duration_seconds:3600
test_duration_actual_seconds:3600.027172
transaction_asset_create_total:2134
transaction_asset_optin_total:1065726
transaction_asset_xfer_total:208325
transaction_asset_close_total:843815
early_average_import_time_sec:8.17
early_cumulative_import_time_sec:1054.50
early_average_imported_tx_per_block:5000.00
early_cumulative_imported_tx_per_block:645000
early_average_block_upload_time_sec:0.36
early_cumulative_block_upload_time_sec:47.54
early_imported_round:130
early_overall_transactions_per_second:611.67
early_uptime_seconds:3600.03
final_average_import_time_sec:8.42
final_cumulative_import_time_sec:3554.10
final_average_imported_tx_per_block:5000.00
final_cumulative_imported_tx_per_block:2110000
final_average_block_upload_time_sec:0.37
final_cumulative_block_upload_time_sec:155.84
final_imported_round:423
final_overall_transactions_per_second:593.68
final_uptime_seconds:3600.03
```

## Test Plan

Run the test runner and observed that the metrics look about right.

### CODE REVIEW

File: util/metrics/metrics.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	prometheus.Register(BlockImportTimeHistogramSeconds)
+	prometheus.Register(CumulativeImportTimeCounter)
+	prometheus.Register(CumulativeTransactionsCounter)
+	prometheus.Register(ImportedTransactionsPerBlockHistogram)
+	prometheus.Register(CurrentRoundGauge)
+}
+
+const (
+
+	// ImportTimePerBlockHistogramName metric name.
```
Comment (tolikzinovyev) [inline util/metrics/metrics.go:17]: I think, if you put a comment right above `const (`, the linter will not require all these silly comments.

File: util/metrics/metrics.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+func RegisterPrometheusMetrics() {
+	prometheus.Register(BlockImportTimeHistogramSeconds)
+	prometheus.Register(CumulativeImportTimeCounter)
+	prometheus.Register(CumulativeTransactionsCounter)
+	prometheus.Register(ImportedTransactionsPerBlockHistogram)
+	prometheus.Register(CurrentRoundGauge)
+}
+
+const (
+
```
Comment (tolikzinovyev) [inline util/metrics/metrics.go:16]: please delete empty line

File: util/metrics/metrics.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	// ImportTimePerBlockHistogramName metric name.
+	ImportTimePerBlockHistogramName = "average_import_time_sec"
+	// ImportTimeCounterName metric name.
+	ImportTimeCounterName = "cumulative_import_time_sec"
+	// TransactionsPerBlockHistogramName metric name.
+	TransactionsPerBlockHistogramName = "average_imported_tx_per_block"
```
Comment (tolikzinovyev) [inline util/metrics/metrics.go:22]: nit: we use `txn` throughout indexer to denote blockchain transactions

File: util/metrics/metrics.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	// CurrentRoundGaugeName metric name.
+	CurrentRoundGaugeName = "current_round"
+)
+
+var (
+	// BlockImportTimeHistogramSeconds average block import duration in seconds.
+	BlockImportTimeHistogramSeconds = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Subsystem: "indexer_daemon",
+			Name:      ImportTimePerBlockHistogramName,
```
Comment (tolikzinovyev) [inline util/metrics/metrics.go:34]: Would be nice to make the name of this variable consistent with the name of the prometheus variable. Right now there are so many variables + string constants, it is confusing.

File: importer/helper.go [hunk 5]
```diff
@@ -259,21 +260,40 @@ func updateAccounting(db idb.IndexerDb, frozenCache map[uint64]bool, filter idb.
...
+			roundStart = now
+			metrics.BlockImportTimeHistogramSeconds.Observe(dt.Seconds())
+			metrics.CumulativeImportTimeCounter.Add(dt.Seconds())
+			metrics.ImportedTransactionsPerBlockHistogram.Observe(float64(txnForRound))
+			metrics.CumulativeTransactionsCounter.Add(float64(txnForRound))
+			metrics.CurrentRoundGauge.Set(float64(currentRound))
+		}
+	}
+	for txn := range txns {
+		fmt.Println("processing a transaction")
```
Comment (tolikzinovyev) [inline importer/helper.go:292]: Are you sure we want to print this for every transaction?

File: util/metrics/metrics.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	// TransactionsPerBlockHistogramName metric name.
+	TransactionsPerBlockHistogramName = "average_imported_tx_per_block"
+	// ImportedTransactionsCounterName metric name.
+	ImportedTransactionsCounterName = "cumulative_imported_tx"
+	// CurrentRoundGaugeName metric name.
+	CurrentRoundGaugeName = "current_round"
+)
+
+var (
+	// BlockImportTimeHistogramSeconds average block import duration in seconds.
```
Comment (tolikzinovyev) [inline util/metrics/metrics.go:30]: Similarly here, a single comment can be on top of `var (`.

File: importer/helper.go [hunk 7]
```diff
@@ -259,21 +260,40 @@ func updateAccounting(db idb.IndexerDb, frozenCache map[uint64]bool, filter idb.
...
+			if txnForRound > 0 {
 				err := db.CommitRoundAccounting(act.RoundUpdates, currentRound, blockHeaderPtr)
 				maybeFail(err, l, "failed to commit round accounting")
 			}
 
+			// record metric
+			now := time.Now()
+			dt := now.Sub(roundStart)
+			roundStart = now
+			metrics.BlockImportTimeHistogramSeconds.Observe(dt.Seconds())
```
Comment (tolikzinovyev) [inline importer/helper.go:284]: Previously we measured import + accounting time, now we measure only accounting time. What is the reason for the change?

Comment (winder) [reply importer/helper.go:284]: Good point, maybe I should move it back.

My thinking was that it's possible for this function to process multiple rounds. I suppose for the tests I'm running (and healthy deployments) that wouldn't happen.

Comment (tolikzinovyev) [reply importer/helper.go:284]: Yes, in a normal scenario we import one block and account one block.

File: util/metrics/metrics.go (hunk 4 above)
Comment (winder) [reply util/metrics/metrics.go]: I tried to shorten these a bit, and make the naming a bit more consistent

File: cmd/block-generator/runner/run.go [hunk 8]
```diff
@@ -82,31 +83,133 @@ func (r *Args) run() error {
...
+}
+
+func getMetric(entry Entry, suffix string, t metricType) (float64, error) {
+	total := 0.0
 	sum := 0.0
 	count := 0.0
+	hasSum := false
+	hasCount := false
 
 	for _, metric := range entry.Data {
```
Comment (tolikzinovyev) [inline cmd/block-generator/runner/run.go:173]: Will `entry.Data` contain same key multiple times? Does this loop select the latest measurement then?

**Review by tolikzinovyev — COMMENTED** [review]
The block generator code seems complicated, and I don't think I understand it fully. I think it can be simplified.

File: cmd/algorand-indexer/daemon.go [hunk 9]
```diff
@@ -197,16 +186,27 @@ func (bih *blockImporterHandler) HandleBlock(block *types.EncodedBlockCert) {
...
 		MaxRound:   uint64(block.Block.Round),
 	}
-	importer.UpdateAccounting(bih.db, bih.cache, filter, logger)
+	rounds, txns := importer.UpdateAccounting(bih.db, bih.cache, filter, logger)
 	dt := time.Now().Sub(start)
-	// record metric
-	importTimeHistogramSeconds.Observe(dt.Seconds())
+
+	// Ignore calls that update >1 round (sneaky migration) and round 0
+	if rounds <= 1 && startRound != 0 {
```
[the remark this answers is not in the document]
Comment (winder) [reply cmd/algorand-indexer/daemon.go:203]: It never had any transactions, so it skews the results

File: util/metrics/metrics.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+		})
+
+	CumulativeImportTime = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Subsystem: "indexer_daemon",
+			Name:      CumulativeImportTimeName,
+			Help:      "Total time in seconds spent uploading and processing blocks.",
+		})
+
+	BlockUploadTime = prometheus.NewSummary(
```
[the remark this answers is not in the document]
Comment (winder) [reply util/metrics/metrics.go:55]: Its a subset of the total import + processing time. So if it takes 8 seconds to it import a round with 20k transactions this might tell us that 1.5 seconds we're spent uploading the raw data

File: cmd/block-generator/runner/run.go [hunk 11]
```diff
@@ -82,31 +83,133 @@ func (r *Args) run() error {
...
-			val := strings.Split(metric, " ")[1]
-			sum, err = strconv.ParseFloat(val, 64)
-		} else if strings.HasPrefix(metric, "indexer_daemon_import_time_sec_count") {
-			val := strings.Split(metric, " ")[1]
-			count, err = strconv.ParseFloat(val, 64)
-		}
-		if err != nil {
-			return fmt.Errorf("unable to parse metric '%s': %w", metric, err)
+
+		if strings.Contains(metric, suffix) {
```
[the remark this answers is not in the document]
Comment (winder) [reply cmd/block-generator/runner/run.go:176]: The Prometheus metrics parser is available as a library. Or Setting up a Prometheus instance and running a proper rate query would also be a more robust solution.

That seemed like overkill for what we're doing so far.

File: cmd/block-generator/runner/run.go (hunk 8 above)
Comment (winder) [reply cmd/block-generator/runner/run.go:173]: There is an array of entry objects and 10 samples are taken while the test runs.

File: cmd/block-generator/runner/run.go [hunk 12]
```diff
@@ -127,11 +230,12 @@ func (r *Args) runTest(indexerURL string, generatorURL string) error {
...
 	for time.Since(start) < r.RunDuration {
 		time.Sleep(r.RunDuration / 10)
-		if err := collector.Collect("indexer_daemon_import_time_sec"); err != nil {
+
+		if err := collector.Collect(metrics.AllMetricNames...); err != nil {
 			return fmt.Errorf("problem collecting metrics: %w", err)
 		}
 	}
-	if err := collector.Collect("indexer_daemon_import_time_sec"); err != nil {
+	if err := collector.Collect(metrics.AllMetricNames...); err != nil {
```
[the remark this answers is not in the document]
Comment (winder) [reply cmd/block-generator/runner/run.go:238]: I'm searching for a substring so a map wouldn't work. The metric prefix should be configurable, we just haven't done it yet. Also, different metrics have different suffixes added automatically.

I don't think we need to do better than this. If we wanted to do better we'd probably replace the whole thing instead of optimizing this code.

File: cmd/block-generator/runner/run.go [hunk 13]
```diff
@@ -82,31 +83,133 @@ func (r *Args) run() error {
...
+		return err
+	}
+
+	importTimeS, err := getMetric(entry, metrics.CumulativeImportTimeName, intTotal)
+	if err != nil {
+		return err
+	}
+	tps := totalTxn / importTimeS
+	tpsKey := "overal_transactions_per_second"
+	msg := fmt.Sprintf("%s_%s:%.2f\n", prefix, tpsKey, tps)
```
[the remark this answers is not in the document]
Comment (winder) [reply cmd/block-generator/runner/run.go:138]: No, that helper function also looks up the metric.

File: util/metrics/metrics.go [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+	prometheus.Register(CumulativeTxns)
+	prometheus.Register(ImportedTxnsPerBlock)
+	prometheus.Register(CurrentRoundGauge)
+	prometheus.Register(CumulativeBlockUploadTime)
+	prometheus.Register(BlockUploadTime)
+}
+
+// Prometheus metric names broken out for reuse.
+const (
+	BlockImportTimeName           = "average_import_time_sec"
```
[the remark this answers is not in the document]
Comment (winder) [reply util/metrics/metrics.go:19]: The histogram comes from taking multiple samples, but whats actually happening with "prometheus.NewSummary" type is an average.

File: util/metrics/metrics.go [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+	BlockUploadTimeName,
+	CumulativeBlockUploadTimeName,
+	ImportedTxnsPerBlockName,
+	CumulativeTxnsName,
+	CurrentRoundGaugeName,
+}
+
+// Initialize the prometheus objects.
+var (
+	BlockImportTimeSeconds = prometheus.NewSummary(
```
[the remark this answers is not in the document]
Comment (winder) [reply util/metrics/metrics.go:41]: I'll add seconds to the other time.

File: cmd/block-generator/runner/run.go [hunk 16]
```diff
@@ -82,33 +83,140 @@ func (r *Args) run() error {
...
+					return sum, nil
+				}
+				if hasTotal {
+					return total, nil
+				}
+			}
+		}
+	}
+
+	return 0.0, fmt.Errorf("metric not found: %s", suffix)
```
Comment (algonautshant) [inline cmd/block-generator/runner/run.go:217]: This may not be necessarily correct. 
Does it matter if the metric is there, but not in the form of rateMetric. And because the request is rateMetric, metric not found is returned. Does this mean "rate metric not found"?

**Review by algonautshant — APPROVED** [review]
Looks great. 
Just a few comments.

File: cmd/algorand-indexer/daemon.go [hunk 17]
```diff
@@ -197,16 +186,23 @@ func (bih *blockImporterHandler) HandleBlock(block *types.EncodedBlockCert) {
...
 	}
-	importer.UpdateAccounting(bih.db, bih.cache, filter, logger)
-	dt := time.Now().Sub(start)
-	// record metric
-	importTimeHistogramSeconds.Observe(dt.Seconds())
+	rounds, txns := importer.UpdateAccounting(bih.db, bih.cache, filter, logger)
+	dt := time.Since(start)
+
+	// Ignore calls that update >1 round (sneaky migration) and round 0 (which is empty)
+	if rounds <= 1 && startRound != 0 {
```
[the remark this answers is not in the document]
Comment (winder) [reply cmd/algorand-indexer/daemon.go:201]: `rounds` is the number of rounds processed in a batch, under normal operation it should be 0 if the round at no transactions, or 1.

Sneaky migration is a way that we can recompute the accounting without reloading transaction/block header data. It's kind of a weird thing, and users shouldn't be doing it, so I felt ok disabling metrics for that case.

File: cmd/block-generator/runner/run.go [hunk 18]
```diff
@@ -82,33 +83,140 @@ func (r *Args) run() error {
...
-			return fmt.Errorf("unable to parse metric '%s': %w", metric, err)
+const (
+	rate metricType = iota
+	intTotal
+	floatTotal
+)
+
+// Helper to record metrics. Supports rates (sum/count) and counters.
+func recordDataToFile(start time.Time, entry Entry, prefix string, out *os.File) error {
+	num := 0
```
[the remark this answers is not in the document]
Comment (winder) [reply cmd/block-generator/runner/run.go:96]: Oops, it isn't. I was using this earlier but not anymore.

File: cmd/block-generator/runner/run.go [hunk 19]
```diff
@@ -82,33 +83,140 @@ func (r *Args) run() error {
...
+	var writeErrors strings.Builder
+	var writeErr error
+	record := func(prefix2, name string, t metricType) {
+		key := fmt.Sprintf("%s%s_%s", prefix, prefix2, name)
+		if err := recordMetricToFile(entry, key, name, t, out); err != nil {
+			writeErr = err
+			if writeErrors.Len() > 0 {
+				writeErrors.WriteString(", ")
+			}
+			writeErrors.WriteString(name)
```
[the remark this answers is not in the document]
Comment (winder) [reply cmd/block-generator/runner/run.go:106]: honestly, the main reason was to avoid having to check return values after every call.

I do kinda like accumulating errors like this, if two of the metrics were missing it would report both of them instead of just the first.

File: cmd/block-generator/runner/run.go (hunk 16 above)
Comment (winder) [reply cmd/block-generator/runner/run.go:217]: I think it's still correct, but you're right about there being some nuance to it... if the metric wasn't a "summary" then the rate fields wouldn't be there and the error could be more precise. On the other hand if the metric isn't there, it would also hit this  error.

File: cmd/block-generator/runner/run.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (winder) [reply cmd/block-generator/runner/run.go:106]: I dunno, I tried changing. Do you think this better?
```
  record := func(prefix2, name string, t metricType) {
    key := fmt.Sprintf("%s%s_%s", prefix, prefix2, name)
    return recordMetricToFile(entry, key, name, t, out)
  }

  if err := record("_average", metrics.BlockImportTimeName, rate); err != nil {    
    return err     
  }    
  if err := record("_cumulative", metrics.BlockImportTimeName, floatTotal); err != nil {    
    return err                                                                                                                                                                                                                                                 
  }                                                                                                                                                                                                                                                            
  if err := record("_average", metrics.ImportedTxnsPerBlockName, rate); err != nil {                                                                                                                                                                           
    return err                                                                                                                                                                                                                                                 
  }                                                                                                                                                                                                                                                            
  if err := record("_cumulative", metrics.ImportedTxnsPerBlockName, intTotal); err != nil {                                                                                                                                                                    
    return err                                                                                                                                                                                                                                                 
  }                                                                                                                                                                                                                                                            
  if err := record("_average", metrics.BlockUploadTimeName, rate); err != nil {                                                                                                                                                                                
    return err                                                                                                                                                                                                                                                 
  }                                                                                                                                                                                                                                                            
  if err := record("_cumulative", metrics.BlockUploadTimeName, floatTotal); err != nil {                                                                                                                                                                       
    return err                                                                                                                                                                                                                                                 
  }                                                                                                                                                                                                                                                            
  if err := record("", metrics.ImportedRoundGaugeName, intTotal); err != nil {                                                                                                                                                                                 
    return err                                                                                                                                                                                                                                                 
  }                                                                                                                                                                                                                                                            
```

[the remark this answers is not in the document]
Comment (algonautshant) [reply cmd/block-generator/runner/run.go:105]: haha maybe not. This looks more like go :-P 
It's your call. I was curious if there was another reason for it. You can leave it as is if you like.
